pdns/dnsname.cc

```diff
@@ -113,72 +113,138 @@ DNSName::DNSName(const char* pos, size_t len, size_t offset, bool uncompress, ui
   packetParser(pos, len, offset, uncompress, qtype, qclass, consumed, 0, minOffset);
 }
 
+static void checkLabelLength(uint8_t length)
+{
+  if (length == 0) {
+    throw std::range_error("no such thing as an empty label to append");
+  }
+  if (length > 63) {
+    throw std::range_error("label too long to append");
+  }
+}
+
+// this parses a DNS name until a compression pointer is found
+size_t DNSName::parsePacketUncompressed(const UnsignedCharView& view, size_t pos, bool uncompress)
+{
+  const size_t initialPos = pos;
+  size_t totalLength = 0;
+  unsigned char labellen = 0;
+
+  do {
+    labellen = view.at(pos);
+    ++pos;
+
+    if (labellen == 0) {
+      --pos;
+      break;
+    }
+
+    if (labellen >= 0xc0) {
+      if (!uncompress) {
+        throw std::range_error("Found compressed label, instructed not to follow");
+      }
+      --pos;
+      break;
+    }
+
+    if ((labellen & 0xc0) != 0) {
+      throw std::range_error("Found an invalid label length in qname (only one of the first two bits is set)");
+    }
+    checkLabelLength(labellen);
+    // reserve one byte for the label length
+    if (totalLength + labellen > s_maxDNSNameLength - 1) {
+      throw std::range_error("name too long to append");
+    }
+    if (pos + labellen >= view.size()) {
+      throw std::range_error("Found an invalid label length in qname");
+    }
+    pos += labellen;
+    totalLength += 1 + labellen;
+  }
+  while (pos < view.size());
+
+  if (totalLength != 0) {
+    auto existingSize = d_storage.size();
+    if (existingSize > 0) {
+      // remove the last label count, we are about to override it */
+      --existingSize;
+      d_storage.resize(existingSize);
```

Comment:
Wondering if this `resize()` is redundant, as it is always followed by another `resize()` call

Reply:
It is indeed redundant, I'll remove it.